discovery-provider/src/tasks/user_replica_set.py

```diff
@@ -0,0 +1,262 @@
+                        cnode_events_lookup[cnode_id]["events"].append(event_type)
+            num_user_replica_set_changes += len(user_events_tx)
+
+    # for each record in user_replica_set_events_lookup, invalidate the old record and add the new record
+    # we do this after all processing has completed so the user record is atomic by block, not tx
+    for user_id, value_obj in user_replica_set_events_lookup.items():
+        logger.info(f"user_replica_set.py | Replica Set Processing Adding {value_obj['user']}")
+        invalidate_old_user(session, user_id)
+        session.add(value_obj["user"])
+
+    for content_node_id, value_obj in cnode_events_lookup.items():
+        logger.info(f"user_replica_set.py | Content Node Processing Adding {value_obj['content_node']}")
+        invalidate_old_cnode_record(session, content_node_id)
+        session.add(value_obj["content_node"])
+
+    return num_user_replica_set_changes, user_ids
+
+# Reconstruct endpoint string from primary and secondary IDs
+# Note that this is BEST EFFORT and may fail, however unlikely
+# In the case of failure where a given user has valid primary/secondaries but no
+# endpoint string, a client must fetch the endpoint associated with ID
+# Indexing CANNOT block on this endpoint string optimization.
+def get_endpoint_string_from_sp_ids(
+        self,
+        update_task,
+        primary,
+        secondaries,
+        redis
+):
+    sp_factory_inst = None
+    endpoint_string = None
+    primary_endpoint = None
+    try:
+        sp_factory_inst, primary_endpoint = get_endpoint_from_id(
+            self,
+            update_task,
+            sp_factory_inst,
+            primary
+        )
+        endpoint_string = "{}".format(primary_endpoint)
+        for secondary_id in secondaries:
+            secondary_endpoint = None
+            sp_factory_inst, secondary_endpoint = get_endpoint_from_id(
+                self,
+                update_task,
+                sp_factory_inst,
+                secondary_id
+            )
+            if secondary_endpoint:
+                endpoint_string = "{},{}".format(endpoint_string, secondary_endpoint)
+            else:
+                logger.info(f"user_replica_set.py | Failed to find secondary info for {secondary_endpoint}")
+    except Exception as exc:
+        logger.error(f"user_replica_set.py | ERROR in get_endpoint_string_from_sp_ids {exc}")
+    logger.info(f"user_replica_set.py | constructed {endpoint_string} from {primary},{secondaries}")
+    return endpoint_string
+
+# Initializes sp_factory if necessary and retrieves spID
+# Returns initialized instance of contract and endpoint
+def get_endpoint_from_id(self, update_task, sp_factory_inst, sp_id):
+    endpoint = None
+    # Get sp_id cache key
+    cache_key = get_sp_id_key(sp_id)
+    # Attempt to fetch from cache
+    sp_info_cached = get_pickled_key(update_task.redis, cache_key)
+    if sp_info_cached:
```

Comment:
from 2/1/21 - move the endpoint reconstruction into the query layer instead of indexing